jax/lax/lax_parallel.py

```diff
@@ -557,13 +557,25 @@ def _all_to_all_translation_rule(c, x, *, split_axis, concat_axis, axis_name,
   replica_groups = _replica_groups(axis_env, axis_name, None)
   if len(replica_groups[0]) == 1:
     return x
-  else:
+  elif platform == 'tpu':
     split_count = len(replica_groups[0])
     if not all(split_count == len(g) for g in replica_groups):
       raise ValueError('Replica groups must be equally sized')
     replica_groups_protos = xc.make_replica_groups(replica_groups)
-    return xops.AllToAll(x, split_axis, concat_axis, split_count,
-                         replica_groups_protos)
+    if concat_axis == split_axis:
+      return xops.AllToAll(x, split_axis, concat_axis, split_count,
+                           replica_groups_protos)
+    else:
+      if concat_axis < split_axis:
+        split_axis += 1
+      elif split_axis < concat_axis:
+        concat_axis += 1
+      x = xla.lower_fun(partial(lax.expand_dims, dimensions=(concat_axis,)), multiple_results=False)(c, x)
+      x = xops.AllToAll(x, split_axis, concat_axis, split_count, replica_groups_protos)
+      x = xla.lower_fun(partial(lax.squeeze, dimensions=(split_axis,)), multiple_results=False)(c, x)
+      return x
+  else:
+    raise NotImplementedError("all_to_all and pswapaxes only supported on TPU")
```

Comment:
Interestingly enough, when I tried running `testAllToAll` on CPU then I was actually getting incorrect results, instead of loud "not supported" crashes, so I figured out it might be a good idea to gate that within JAX. My other patch fixes this limitation by adding the fallback through `all_gather`, but I need to land this one first or else the TPU tests start failing.

Reply:
Should we leave a TODO here, and maybe a corresponding bug report / feature request for the XLA:CPU/GPU team? It may be quick for them to fix the issue, so we may not need the all_gather fallback at the JAX level for long.

Reply:
Sure, I'll create one. I wanted to submit the CPU bug anyway

Reply:
apologies for restarting an old thread (should I rather open a new one) but is there any update on the cpu support for all_to_all? (it seems that since this discussion the GPU support has been added)